rfcs/text/0011_global_search.md

```diff
@@ -0,0 +1,693 @@
+ * Currently empty and only present to keep the API future-proof.
+ */
+interface GlobalSearchOptions {}
+
+/**
+ * Representation of a result returned by a {@linlkGlobalSearchResultProvider | result provider}
+ */
+interface GlobalSearchResult {
+  /** an id that should be unique for an individual provider's results */
+  id: string;
+  /** the title/label of the result */
+  title: string;
+  /** the type of result */
+  type: GlobalSearchResultType;
+  /** an optional EUI icon name to associate with the search result. If not specified, will use a default icon */
+  icon?: string;
+  /** The url to navigate to this result. This can be either an absolute url, or a path relative to the performing request's basePath */
+  url: string;
+  /** the score of the result, from 1 (lowest) to 100 (highest) */
+  score: number;
+  /** an optional record of metadata for this result */
+  meta?: Record<string, any>;
+}
+```
+
+#### server
+
+```ts
+/**
+ * Context passed to server-side {@GlobalSearchResultProvider | result provider}'s `find` method.
+ */
+export interface GlobalSearchProviderContext {
+  request: KibanaRequest;
+  core: {
+    savedObjects: {
+      client: SavedObjectsClientContract;
+      typeRegistry: ISavedObjectTypeRegistry;
+    };
+    elasticsearch: {
+      legacy: {
+        client: IScopedClusterClient;
+      };
+    };
+    uiSettings: {
+      client: IUiSettingsClient;
+    };
+  };
+}
+
+/**
+ * GlobalSearch result provider, to be registered using the {@link GlobalSearchSetup | global search API}
+ */
+type GlobalSearchResultProvider = {
+  id: string;
+  find(
+    term: string,
+    options: GlobalSearchOptions,
+    context: GlobalSearchProviderContext
+  ): Observable<GlobalSearchResult[]>;
+};
+```
+
+Notes: 
+- initial implementation will only provide a static / non extensible `GlobalSearchProviderContext` context.
+It would be possible to allow plugins to register their own context providers as it's done for `RequestHandlerContext`,
+but this will not be done until the need arises.
+- Because of the previous point, the performing `request` object is also exposed on the context to allow result providers
+to scope their custom services if needed.
+
+#### public
+
+```ts
+/**
+ * GlobalSearch result provider, to be registered using the {@link GlobalSearchSetup | global search API}
+ */
+type GlobalSearchResultProvider = {
+  id: string;
+  find(
+    term: string,
+    options: GlobalSearchOptions,
+  ): Observable<GlobalSearchResult[]>;
+};
+```
+
+Notes: 
+- The client-side version of `GlobalSearchResultProvider` is slightly difference than the 
+server one, as there is no `context` parameter on the `find` signature.
+
+### Plugin API
+
+#### common types
+
+```ts
+/**
+ * Response returned from the {@link GlobalSearchServiceStart | global search service}'s `find` API
+ */
+type GlobalSearchResponse<ResultType extends GlobalSearchResult = GlobalSearchResult> = {
```

Comment:
Do we get any benefit in sharing this type between the client and the server? We could always just have separate types for the client and the server.

Reply:
The goal here was more to show how the server/client parts were similar. No problem with using distinct non generic types instead in the final implementation. Will update.

Reply:
Done in https://github.com/elastic/kibana/pull/64284/commits/5d750e072f4606634ff7e7c3c234c00fc0fe46ba